### PR TITLE
Improvements of the .NET Core build

### PR DESCRIPTION
**Pull request by bkragl**
**Edit:** This PR evolved quite a bit from its original version. Points 1 and 3 below do not apply anymore, and point 5 was added.

This PR makes the following (in my opinion) improvements:

~1. The build output of the BoogieDriver project is set to the `Binaries` directory, as for the old build. Thus the compiled executable is written to `Binaries/Boogie.dll`. The option `AppendTargetFrameworkToOutputPath` is set to `false` because otherwise the location would depend on the target framework (e.g., `Binaries/netcoreapp2.2/Boogie.dll`). Also, the created NuGet package is placed into `Binaries/nupkg/`.~
2. The lit test tool can now be called with `-D dotnet` to run the regression tests with the .NET Core build.
~3. A wrapper script `Binaries/boogie-dotnet` is added. This makes it more convenient to call on the command line or to configure the Emacs Flycheck mode for Boogie.~
4. The build instructions in `README.md` are cleaned up and emphasize that .NET Core is from now on the preferred way to build/run Boogie. I think this was the consensus from previous discussions. Also, more details for the three supported SMT solvers are given.
5. **Added**: The Travis CI build is switched from Mono to .NET Core and configured to automatically deploy tagged releases.

@michael-emmi, please let me know if you agree with the above changes, or if anything should be updated.

### CODE REVIEW

**Comment by michael-emmi** [comment]
Just curious, why the wrapper script and all that?

According to [the docs](https://docs.microsoft.com/en-us/dotnet/core/tools/global-tools-how-to-create#setup-the-global-tool) you can just install the tool from a package:
```console
dotnet tool install --global --add-source ./nupkg botsay
```
i.e., the package does not need to exist (yet) in the online NuGet repository.

Also, it seems that local tools are now possible too, avoiding a global install: https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-tool-install

But maybe creating a script is preferable for some reason?

File: Source/BoogieDriver/BoogieDriver-NetCore.csproj [hunk 1]
```diff
@@ -9,12 +9,14 @@
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
-    <AssemblyName>BoogieDriver</AssemblyName>
+    <AssemblyName>Boogie</AssemblyName>
```
Comment (michael-emmi) [inline Source/BoogieDriver/BoogieDriver-NetCore.csproj]: Why this change? Maybe better to maintain the legacy structure as much as possible?

Comment (bkragl) [reply Source/BoogieDriver/BoogieDriver-NetCore.csproj]: I thought it's nice to call the "main" executable DLL just `Boogie.dll`, but I understand the desire to name it the same as the project. I'll rename it back to `BoogieDriver`.
But you agree with `Binaries/` in the repository root as the output path?

**Comment by bkragl** [comment]
The wrapper script is intended for developers. During (iterative) development and testing it does not seem right to always do an installation step.

But we should add instructions on how to install the global tool from nuget.org for people who are just interested in using Boogie.

File: Source/BoogieDriver/BoogieDriver-NetCore.csproj (hunk 1 above)
Comment (michael-emmi) [reply Source/BoogieDriver/BoogieDriver-NetCore.csproj]: Well, since that is how Boogie was built before the `-NetCore.csproj` variations, yes.

**Comment by michael-emmi** [comment]
> During (iterative) development and testing it does not seem right to always do an installation step.

Agreed, though the step I suggest would be an “install” that is *local* to the workspace, i.e., after generating `BoogieDriver.dll` (which is anyhow done in iterative development) generate the package then have dotnet “install” the tool in the workspace directory, which the tests can use — I’m supposing those steps don’t add significant overhead to iterative development.

I think either approach could work, and I suggested the above to avoid having a script that could be generated automatically, and avoid divergence with what’s actually on regular (global) installs.

Anyhow I won’t insist on this point, since I’m not so frequently involved in the Boogie development cycle :-)

**Comment by michael-emmi** [comment]
@bkragl how about updating `.travis.yml` to build for dotnet core (also)?

If you do that, then we can almost immediately automate publishing versioned dotnet global tools to NuGet.org with #174.

**Comment by bkragl** [comment]
Sure, we can add a .NET Core build to Travis.

However, I was rethinking my proposed updates in this PR. Specifically, how to run Boogie as a developer and from the Lit tests. I don't particularly like the output of the BoogieDriver project to be written to `./Binaries/` and having that script (actually I would hope that `Binaries` can be deleted at some point), but the installation you mentioned above is also not completely satisfactory.

I tried `dotnet tool install` and found that there are three options where exactly one can be given:

* `--local`: This is meant for tools that should be installed and available "locally". It relies on a tool manifest file that specifies the desired tools and versions. However, the actual installation is not done locally, but somewhere in the HOME directory (such that it can be shared in case different projects want to have the same local tool). According to my understanding this would be a perfect use case for Coco/R for us. My plan is to create a Coco/R package on nuget.org that we can include as a local tool into Boogie.

* `--global`: This makes a global installation for the current user. This would be the right thing for a user of Boogie, but not for development/testing.

* `--tool-path`: This takes a directory as target for the installation (which would be truly local). Here we could say `--tool-path test-installation`. However, it seems that there is no way to explicitly tell dotnet that the installation should be performed from the local package. You have to specify `--add-source Source/BoogieDriver/bin/$CONFIGURATION`, but I'm not sure if there could be ambiguity if there is already a published version on nuget.org with the same version number (because during development the version number is not yet bumped).

Do you think we should try the last option, or for now call `dotnet Source/BoogieDriver/bin/Debug/netcoreapp2.2/BoogieDriver.dll` for testing?

**Comment by bkragl** [comment]
BTW, we should perhaps upgrade to .NET Core 3.1 which is a long term support (LTS) release.

**Comment by michael-emmi** [comment]
> Do you think we should try the last option, or for now call dotnet Source/BoogieDriver/bin/Debug/netcoreapp2.2/BoogieDriver.dll for testing?

The would avoid ad-host invocation scripts, as long as the potential caveat you mention can be resolved:

> but I'm not sure if there could be ambiguity if there is already a published version on nuget.org with the same version number

I’m supposing that the local installation would take precedence? Is there an easy way to verify this?

**Comment by michael-emmi** [comment]
> BTW, we should perhaps upgrade to .NET Core 3.1 which is a long term support (LTS) release.

Absolutely.

**Comment by michael-emmi** [comment]
Hey @bkragl just noticed something

> A wrapper script Binaries/boogie-dotnet is added. This makes it more convenient to call on the command line or to configure the Emacs Flycheck mode for Boogie.

The current build seems to generate `./Source/BoogieDriver/bin/Debug/netcoreapp3.1/BoogieDriver`, which can be invoked directly.

**Comment by michael-emmi** [comment]
Also see #180 for improvements to the -NetCore.csproj files themselves.

**Comment by michael-emmi** [comment]
@bkragl how’s this one coming? need any assist?

**Comment by bkragl** [comment]
> @bkragl how’s this one coming? need any assist?

I'll post a proposal soon.

**Comment by bkragl** [comment]
@michael-emmi: I basically restarted this PR, pulling in the recent changes we did. I switched the Travis build to .NET Core and added configurations to deploy to nuget.org and GitHub releases (we'll have to see if this works as intended). Unfortunately there seems to be a problem during compilation related to GitVersion.

This is the error message I saw in the build log:
```
/home/travis/.nuget/packages/gitversiontask/5.1.2/build/GitVersionTask.targets(10,9): error : InvalidOperationException: Could not find a 'develop' or 'master' branch, neither locally nor remotely. [/home/travis/build/boogie-org/boogie/Source/CodeContractsExtender/CodeContractsExtender-NetCore.csproj]
```

Perhaps the git metadata is not available in the Travis environment? Could you look into that?

If this is fixed, we also have to add a GitHub oauth token, if we want to have files attached to the releases on GitHub.

**Comment by michael-emmi** [comment]
> Perhaps the git metadata is not available in the Travis environment? Could you look into that?

Looks like you have to update increase the clone depth:
https://docs.travis-ci.com/user/customizing-the-build#git-clone-depth

File: Source/BoogieDriver/BoogieDriver-NetCore.csproj [hunk 2]
```diff
@@ -13,7 +13,6 @@
     <AssemblyName>BoogieDriver</AssemblyName>
     <TargetFramework>netcoreapp3.1</TargetFramework>
     <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
-    <PackageOutputPath>nupkg</PackageOutputPath>
```
Comment (michael-emmi) [inline Source/BoogieDriver/BoogieDriver-NetCore.csproj]: Why are we deleting this?

File: .travis.yml [hunk 3]
```diff
@@ -1,27 +1,44 @@
...
-  # Run driver tests
-  - ln -s $(find $PWD/z3* -name z3 -type f) Binaries/z3.exe
-  - lit -v Test/
+  - dotnet build -c ${CONFIGURATION} ${SOLUTION}
+  - dotnet test --no-build -c ${CONFIGURATION} ${SOLUTION}
+  - lit -v -D dotnet -D configuration=${CONFIGURATION} Test
+deploy:
+  - provider: script
+    script:
+      - dotnet nuget push Source/BoogieDriver/bin/${CONFIGURATION}/Boogie*.nupkg -k ${NUGET_API_KEY} -s https://api.nuget.org/v3/index.json
```
Comment (michael-emmi) [inline .travis.yml]: Wouldn't the path `Source/BoogieDriver/nupkg/Boogie.2.4.2.nupkg` be better here?

File: Source/BoogieDriver/BoogieDriver-NetCore.csproj (hunk 2 above)
Comment (bkragl) [reply Source/BoogieDriver/BoogieDriver-NetCore.csproj]: I wanted to see where the package is placed by default and found it cleaner to have no overlap between the build output of different configurations, i.e., everything concerning a particular configuration is placed in `bin/${CONFIGURATION}/`. I'll add it back if you prefer.

Comment (michael-emmi) [reply Source/BoogieDriver/BoogieDriver-NetCore.csproj]: We’re considering builds with multiple configurations?

Comment (bkragl) [reply Source/BoogieDriver/BoogieDriver-NetCore.csproj]: Yes. Currently Travis is configured to build and test the Debug and Release configurations. The deployment should only happen for Release.

Comment (michael-emmi) [reply Source/BoogieDriver/BoogieDriver-NetCore.csproj]: But those are two separate builds, one configuration per build.

Wouldn’t it be nice to find the package in the same place independently of configuration?

Comment (bkragl) [reply Source/BoogieDriver/BoogieDriver-NetCore.csproj]: I was thinking of a local clone of the repository. The same argument could be made for the other build output. Putting everything into `bin` instead of `bin/Debug/netcore3.1` would make it easier to find the binaries (e.g., in test scripts). Still, I'm against that because it might be useful to have multiple configurations compiled simultaneously.

That being said, for the package I don't care.

Comment (michael-emmi) [reply Source/BoogieDriver/BoogieDriver-NetCore.csproj]: Right. For `bin` I agree, because, like you say, you might want to build both locally.

But these two lines are just about the package and related configuration for Travis, which only builds one configuration, and can be simplified accordingly.

Comment (michael-emmi) [reply Source/BoogieDriver/BoogieDriver-NetCore.csproj]: Then again, maybe deleting the line above is better than simplifying the line below.

Let’s keep your version :-).

File: .travis.yml (hunk 3 above)
Comment (bkragl) [reply .travis.yml]: Any idea for a clean way to get rid of the wildcard?

Comment (michael-emmi) [reply .travis.yml]: No, but this seems clean already. Are you worried that this could match multiple files in a single travis build?

Comment (bkragl) [reply .travis.yml]: It's more about knowing that there is a single file and still using a wildcard. But you are right, this is probably more stable than some way of trying to compute the exact file name. It's similar to the code that puts Z3 into PATH. We won't have to change it when we switch to a newer version.

File: .travis.yml [hunk 4]
```diff
@@ -1,27 +1,44 @@
...
+    skip_cleanup: true
+    on:
+      branch: master
+      tags: true
+      condition: ${CONFIGURATION}=Release
+  - provider: releases
+    api_key: TODO
```
Comment (bkragl) [inline .travis.yml]: @michael-emmi, do you know how to fill this? I can look into it, but I'm not sure if I have the necessary privileges since I'm not an owner of the Boogie repo.

Or do we want to eliminate the deployment to GitHub?

Comment (michael-emmi) [reply .travis.yml]: I’m not clear on the utility of GitHub Releases given NuGet releases, but I’ve added an API key anyhow :-)

**Comment by michael-emmi** [comment]
@bkragl is this ready to merge?

**Comment by bkragl** [comment]
The readme file should be updated with a pointer to the NuGet package for people that are only interested in using Boogie, and some more compilation instructions for .NET Core. I can do that tomorrow.

Otherwise I think we can merge.

**Comment by michael-emmi** [comment]
@bkragl cool let me know when you add those and I can merge when ready.

**Comment by bkragl** [comment]
> @bkragl cool let me know when you add those and I can merge when ready.

Done. I also updated the original description of this PR.
